sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -119,9 +122,27 @@ object ProfilerOnExecutor extends Logging {
       val stageId = taskCtx.stageId
       if (stageRanges.contains(stageId)) {
         synchronized {
-          activeStages.add(taskCtx.stageId)
-          enable()
-          startPollingDriver()
+          if (taskLimit <= 0) {
+            // Unlimited tasks per stage
+            activeStages.add(taskCtx.stageId)
+            enable()
+            startPollingDriver()
+          } else {
+            // Limited tasks per stage
+            if (stageTaskCount.getOrElse(stageId, 0) < taskLimit) {
+              activeStages.add(taskCtx.stageId)
+              enable()
+              startPollingDriver()
+            }
+            taskCtx.addTaskCompletionListener[Unit] { _ =>
+              val currentCount = stageTaskCount.getOrElse(stageId, 0)
+              if (currentCount < taskLimit) {
+                stageTaskCount(stageId) = currentCount + 1
```

Comment:
This is not multithread safe.  Spark guarantees the listener callbacks of a task are called serially, but makes no such guarantee across tasks.

Reply:
thanks, added a synchronized around it.